File: 0783-minimum-distance-between-bst-nodes/0783-minimum-distance-between-bst-nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function(root) {
    if (!root) return 0;
    var minDiff = Infinity;
    var stack = [root];
    var vals = [];
    while (stack.length) {
        let node = stack.pop();
        vals.push(node.val)
        if (node.left) {
            stack.push(node.left)
        }
        if (node.right) {
            stack.push(node.right)
        }
    }
    var sortVal = vals.sort((a, b) => a - b);
    for (var i = 0; i < sortVal.length - 1; i++) {
        minDiff = Math.min(minDiff, sortVal[i + 1] - sortVal[i])
    }
    return minDiff;
};

File: 0783-minimum-distance-between-bst-nodes/0783-minimum-distance-between-bst-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        self.ans = float('inf')
        self.pred = None
        self.inOrder(root)
        return self.ans
    
    
    def inOrder(self, node: TreeNode) -> None:
        if not node: return
        
        self.inOrder(node.left)
        if self.pred is not None:
            self.ans = min(self.ans, node.val - self.pred)
        self.pred = node.val
        self.inOrder(node.right)
        